src/snowflake/snowpark/_internal/server_connection.py

```diff
@@ -420,7 +425,13 @@ def notify_query_listeners(self, query_record: QueryRecord) -> None:
     def execute_and_notify_query_listener(
         self, query: str, **kwargs: Any
     ) -> SnowflakeCursor:
-        results_cursor = self._cursor.execute(query, **kwargs)
+        try:
+            results_cursor = self._cursor.execute(query, **kwargs)
+        except Error as err:
+            self.notify_query_listeners(
+                QueryRecord(err.sfqid, err.query), is_error=True
+            )
+            raise err
```

Comment:
this pattern is repeated thrice. I'm wondering if we can do something to remove duplication.